File: output/deepseek_r1/avatar/direct_translation/python/java/atcoder_ABC122_D.java

import java.util.HashMap;
import java.util.Scanner;

public class atcoder_ABC122_D {
    static int n;
    static int mod = (int) 1e9 + 7;
    static HashMap<String, Integer>[] memo;

    public static boolean ok(String last4) {
        for (int i = 0; i < 4; i++) {
            char[] arr = last4.toCharArray();
            if (i >= 1) {
                char temp = arr[i];
                arr[i] = arr[i - 1];
                arr[i - 1] = temp;
            }
            String s = new String(arr);
            if (s.contains("AGC")) {
                return false;
            }
        }
        return true;
    }

    public static int dfs(int cur, String last3) {
        if (memo[cur].containsKey(last3)) {
            return memo[cur].get(last3);
        }
        if (cur == n) {
            return 1;
        }
        int ret = 0;
        char[] chars = {'A', 'C', 'G', 'T'};
        for (char c : chars) {
            String new4 = last3 + c;
            if (ok(new4)) {
                String newLast3 = new4.substring(1);
                ret = (ret + dfs(cur + 1, newLast3)) % mod;
            }
        }
        memo[cur].put(last3, ret);
        return ret;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        n = scanner.nextInt();
        memo = (HashMap<String, Integer>[]) new HashMap<?, ?>[n + 1];
        for (int i = 0; i <= n; i++) {
            memo[i] = new HashMap<>();
        }
        System.out.println(dfs(0, "TTT"));
    }
}

File: output/deepseek_r1/avatar/direct_translation/java/python/atcoder_ABC122_D.py

class Solution:
    def __init__(self):
        self.N = int(input())
        self.MOD = 10**9 + 7
        self.TOKENS = ['A', 'C', 'G', 'T']
        self.memo = [{} for _ in range(self.N + 1)]

    def calc(self):
        return self.dfs(0, 'TTT')

    def isOK(self, last4):
        if 'AGC' in last4:
            return False
        for i in range(3):
            chars = list(last4)
            chars[i], chars[i+1] = chars[i+1], chars[i]
            if 'AGC' in ''.join(chars):
                return False
        return True

    def dfs(self, current, last3):
        if last3 in self.memo[current]:
            return self.memo[current][last3]
        if current == self.N:
            return 1
        res = 0
        for c in self.TOKENS:
            new_last4 = last3 + c
            if self.isOK(new_last4):
                next_last3 = last3[1:] + c
                res = (res + self.dfs(current + 1, next_last3)) % self.MOD
        self.memo[current][last3] = res
        return res

if __name__ == '__main__':
    ins = Solution()
    print(ins.calc())

